salt/modules/dockerng.py

```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+        __opts__,
+        grains,
+        # Not sure if these two are correct
+        __opts__['id'],
+        __opts__['environment'],
+        pillar=pillar_override,
+        pillarenv=pillarenv
+    )
+    ret = pillar.compile_pillar()
+    if pillar_override and isinstance(pillar_override, dict):
+        ret.update(pillar_override)
+    return ret
+
+
+def call(name, function, *args, **kwargs):
+    '''
+    Executes a salt function inside a container
+
+    .. code-block:: bash
+
+        salt myminion dockerng.call test.ping
+
+        salt myminion test.arg arg1 arg2 key1=val1
+
+    The container does not need to have Salt installed, but Python
+    is required.
+
+    .. versionadded:: Carbon
+
+    '''
+    # put_archive reqires the path to exist
+    ret = __salt__['dockerng.run_all'](name, 'mkdir -p /tmp/salt_thin')
+    if ret['retcode'] != 0:
+        return {'result': False, 'comment': ret['stderr']}
+
+    if function is None:
+        raise CommandExecutionError('Missing function parameter')
+
+    # move salt into the container
+    thin_path = salt.utils.thin.gen_thin(__opts__['cachedir'])
+    with io.open(thin_path, 'rb') as file:
+        _client_wrapper('put_archive', name, '/tmp/salt_thin', file)
+
+    salt_argv = [
+        'python',
+        '/tmp/salt_thin/salt-call',
+        '--retcode-passthrough',
+        '--local',
+        '--out', 'json',
+        '-l', 'quiet',
+        '--',
+        function
+    ] + list(args) + ['{0}={1}'.format(key, value) for (key, value) in kwargs.items() if not key.startswith('__')]
+
+    ret = __salt__['dockerng.run_all'](name,
+                                       list2cmdline(map(str, salt_argv)))
+    # python not found
+    if ret['retcode'] == 127:
+        raise CommandExecutionError(ret['stderr'])
+    elif ret['retcode'] != 0:
+        return {'result': False, 'comment': ret['stderr']}
+
+    # process "real" result in stdout
+    try:
+        data = salt.utils.find_json(ret['stdout'])
+        return data.get('local', data)
+    except ValueError:
+        return {'result': False,
+                'comment': 'Can\'t parse container command output'}
+
+
+def sls(name, mods=None, saltenv='base', **kwargs):
+    '''
+    Apply the highstate defined by the specified modules.
+
+    For example, if your master defines the states ``web`` and ``rails``, you
+    can apply them to a container:
+    states by doing:
+
+    .. code-block:: bash
+
+        salt myminion dockerng.sls compassionate_mirzakhani mods=rails,web
+
+    The container does not need to have Salt installed, but Python
+    is required.
+
+    .. versionadded:: Carbon
+    '''
+    if mods is not None:
+        mods = [x.strip() for x in mods.split(',')]
+    else:
+        mods = []
+
+    # gather grains from the container
+    grains = __salt__['dockerng.call'](name, function='grains.items')
+
+    # compile pillar with container grains
+    pillar = _gather_pillar(saltenv, {}, grains)
```

Comment:
I think this call is wrong now that I changed the signature. I think I have to pass `**grains`here to convert the dict to keyword arguments.